File: objectModel/Java/objectmodel/src/test/java/com/microsoft/commondatamodel/objectmodel/cdm/relationship/RelationshipTest.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.cdm.relationship;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.microsoft.commondatamodel.objectmodel.TestHelper;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmE2ERelationship;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmEntityDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmManifestDeclarationDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmManifestDefinition;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmRelationshipDiscoveryStyle;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.E2ERelationship;
import com.microsoft.commondatamodel.objectmodel.storage.LocalAdapter;
import com.microsoft.commondatamodel.objectmodel.utilities.JMapper;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import org.testng.Assert;
import org.testng.annotations.Test;

public class RelationshipTest {
  private static final String CDM = "cdm";
  private static final String TESTS_SUBPATH = new File(CDM, "relationship").toString();

  /**
   * Testing calculation of relationships and that those relationships are
   * properly added to manifest objects
   */
  @Test
  public void testCalculateRelationshipsAndPopulateManifests() throws IOException, InterruptedException {
    final CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "testCalculateRelationshipsAndPopulateManifests", null);
    final CdmManifestDefinition rootManifest = corpus
        .<CdmManifestDefinition>fetchObjectAsync("local:/default.manifest.cdm.json").join();
    final CdmManifestDefinition subManifest = corpus
        .<CdmManifestDefinition>fetchObjectAsync(rootManifest.getSubManifests().get(0).getDefinition()).join();

    corpus.calculateEntityGraphAsync(rootManifest).join();
    rootManifest.populateManifestRelationshipsAsync().join();

    final List<E2ERelationship> expectedAllManifestRels = JMapper.MAP.readValue(
        TestHelper.getExpectedOutputFileContent(TESTS_SUBPATH, "testCalculateRelationshipsAndPopulateManifests",
            "expectedAllManifestRels.json"),
        new TypeReference<List<E2ERelationship>>() {
        });
    final List<E2ERelationship> expectedAllSubManifestRels = JMapper.MAP.readValue(
        TestHelper.getExpectedOutputFileContent(TESTS_SUBPATH, "testCalculateRelationshipsAndPopulateManifests",
            "expectedAllSubManifestRels.json"),
        new TypeReference<List<E2ERelationship>>() {
        });

    // check that each relationship has been created correctly
    verifyRelationships(rootManifest, expectedAllManifestRels);
    verifyRelationships(subManifest, expectedAllSubManifestRels);
  }

  /**
   * Testing calculation of relationships and that those relationships are
   * properly added to manifest objects.
   */
  @Test
  public void TestCalculateRelationshipsAndPopulateManifestsWithExclusiveFlag()
      throws InterruptedException, ExecutionException, IOException {
    final CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "testCalculateRelationshipsAndPopulateManifests", null);

    final CdmManifestDefinition rootManifest = corpus
        .<CdmManifestDefinition>fetchObjectAsync("local:/default.manifest.cdm.json").get();
    final CdmManifestDefinition subManifest = corpus
        .<CdmManifestDefinition>fetchObjectAsync(rootManifest.getSubManifests().get(0).getDefinition()).get();

    corpus.calculateEntityGraphAsync(rootManifest).join();
    rootManifest.populateManifestRelationshipsAsync(CdmRelationshipDiscoveryStyle.Exclusive).join();

    final List<E2ERelationship> expectedExclusiveManifestRels = JMapper.MAP.readValue(
        TestHelper.getExpectedOutputFileContent(TESTS_SUBPATH, "TestCalculateRelationshipsAndPopulateManifests",
            "expectedExclusiveManifestRels.json"),
        new TypeReference<List<E2ERelationship>>() {
        });
    final List<E2ERelationship> expectedExclusiveSubManifestRels = JMapper.MAP.readValue(
        TestHelper.getExpectedOutputFileContent(TESTS_SUBPATH, "TestCalculateRelationshipsAndPopulateManifests",
            "expectedExclusiveSubManifestRels.json"),
        new TypeReference<List<E2ERelationship>>() {
        });

    // check that each relationship has been created correctly
    verifyRelationships(rootManifest, expectedExclusiveManifestRels);
    verifyRelationships(subManifest, expectedExclusiveSubManifestRels);
  }

  /**
   * Testing calculation of relationships and that those relationships are
   * properly added to manifest objects setting the populate flag to None.
   */
  @Test
  public void testCalculateRelationshipsAndPopulateManifestsWithNoneFlag()
      throws ExecutionException, InterruptedException {
    final CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "testCalculateRelationshipsAndPopulateManifests", null);

    final CdmManifestDefinition rootManifest = corpus
        .<CdmManifestDefinition>fetchObjectAsync("local:/default.manifest.cdm.json").get();
    final CdmManifestDefinition subManifest = corpus
        .<CdmManifestDefinition>fetchObjectAsync(rootManifest.getSubManifests().get(0).getDefinition()).get();

    corpus.calculateEntityGraphAsync(rootManifest).join();
    // make sure no relationships are added when "none" relationship option is passed in
    rootManifest.populateManifestRelationshipsAsync(CdmRelationshipDiscoveryStyle.None).join();

    Assert.assertEquals(0, rootManifest.getRelationships().getCount());
    Assert.assertEquals(0, subManifest.getRelationships().getCount());
  }

  /**
   * Testing calculation of relationships when resolved entities are listed in the manifest
   */
  @Test
  public void testCalculateRelationshipsOnResolvedEntities()
      throws JsonMappingException, JsonProcessingException, IOException, InterruptedException {
    final List<E2ERelationship> expectedResolvedManifestRels = 
      JMapper.MAP.readValue(TestHelper.getExpectedOutputFileContent(
      TESTS_SUBPATH,
      "testCalculateRelationshipsOnResolvedEntities", 
      "expectedResolvedManifestRels.json"),
      new TypeReference<List<E2ERelationship>>() {
      });
    final List<E2ERelationship> expectedResolvedSubManifestRels = 
      JMapper.MAP.readValue(TestHelper.getExpectedOutputFileContent(
      TESTS_SUBPATH,
      "testCalculateRelationshipsOnResolvedEntities",
      "expectedResolvedSubManifestRels.json"),
      new TypeReference<List<E2ERelationship>>() {
      });
    final List<E2ERelationship> expectedResolvedExcManifestRels = 
      JMapper.MAP.readValue(TestHelper.getExpectedOutputFileContent(
      TESTS_SUBPATH,
      "testCalculateRelationshipsOnResolvedEntities", 
      "expectedResolvedExcManifestRels.json"),
      new TypeReference<List<E2ERelationship>>() {
      });
    final List<E2ERelationship> expectedResolvedExcSubManifestRels = 
      JMapper.MAP.readValue(TestHelper.getExpectedOutputFileContent(
      TESTS_SUBPATH,
      "testCalculateRelationshipsOnResolvedEntities",
      "expectedResolvedExcSubManifestRels.json"),
      new TypeReference<List<E2ERelationship>>() {
      });
    final String testInputPath = TestHelper.getInputFolderPath(TESTS_SUBPATH, "TestCalculateRelationshipsOnResolvedEntities");
    final CdmCorpusDefinition corpus = new CdmCorpusDefinition();
    corpus.getStorage().mount("local", new LocalAdapter(testInputPath));
    corpus.getStorage().setDefaultNamespace("local");
    final CdmManifestDefinition rootManifest = corpus.<CdmManifestDefinition>fetchObjectAsync("local:/default.manifest.cdm.json").join();
    final CdmManifestDefinition resolvedManifest = loadAndResolveManifest(corpus, rootManifest, "-resolved");
    String subManifestPath = corpus.getStorage().createAbsoluteCorpusPath(resolvedManifest.getSubManifests().getAllItems().get(0).getDefinition());
    CdmManifestDefinition subManifest = corpus.<CdmManifestDefinition>fetchObjectAsync(subManifestPath).join();

    // using createResolvedManifest will only populate exclusive relationships
    verifyRelationships(resolvedManifest, expectedResolvedExcManifestRels);
    verifyRelationships(subManifest, expectedResolvedExcSubManifestRels);

    // check that each relationship has been created correctly with the all flag
    resolvedManifest.populateManifestRelationshipsAsync().join();
    subManifest.populateManifestRelationshipsAsync().join();
    verifyRelationships(resolvedManifest, expectedResolvedManifestRels);
    verifyRelationships(subManifest, expectedResolvedSubManifestRels);

    // it is not enough to check if the relationships are correct.
    // We need to check if the incoming and outgoing relationships are
    // correct as well. One being correct can cover up the other being wrong
    // A
    final CdmEntityDefinition aEnt = corpus.<CdmEntityDefinition>fetchObjectAsync(resolvedManifest.getEntities().getAllItems().get(0).getEntityPath(), resolvedManifest).join();
    final ArrayList<CdmE2ERelationship> aInRels = corpus.fetchIncomingRelationships(aEnt);
    final ArrayList<CdmE2ERelationship> aOutRels = corpus.fetchOutgoingRelationships(aEnt);
    Assert.assertEquals(aInRels.size(), 0);
    Assert.assertEquals(aOutRels.size(), 1);
    Assert.assertEquals(aOutRels.get(0).getFromEntity(), "local:/A-resolved.cdm.json/A");
    Assert.assertEquals(aOutRels.get(0).getToEntity(), "local:/B-resolved.cdm.json/B");

    // B
    final CdmEntityDefinition bEnt = corpus.<CdmEntityDefinition>fetchObjectAsync(resolvedManifest.getEntities().getAllItems().get(1).getEntityPath(), resolvedManifest).join();
    final ArrayList<CdmE2ERelationship> bInRels = corpus.fetchIncomingRelationships(bEnt);
    final ArrayList<CdmE2ERelationship> bOutRels = corpus.fetchOutgoingRelationships(bEnt);
    Assert.assertEquals(bInRels.size(), 2);
    Assert.assertEquals(bInRels.get(0).getFromEntity(), "local:/A-resolved.cdm.json/A");
    Assert.assertEquals(bInRels.get(0).getToEntity(), "local:/B-resolved.cdm.json/B");
    Assert.assertEquals(bInRels.get(1).getFromEntity(), "local:/sub/C-resolved.cdm.json/C");
    Assert.assertEquals(bInRels.get(1).getToEntity(), "local:/B-resolved.cdm.json/B");
    Assert.assertEquals(bOutRels.size(), 0);

    // C
    final CdmEntityDefinition cEnt = corpus.<CdmEntityDefinition>fetchObjectAsync(subManifest.getEntities().getAllItems().get(0).getEntityPath(), subManifest).join();
    final ArrayList<CdmE2ERelationship> cInRels = corpus.fetchIncomingRelationships(cEnt);
    final ArrayList<CdmE2ERelationship> cOutRels = corpus.fetchOutgoingRelationships(cEnt);
    Assert.assertEquals(cInRels.size(), 0);
    Assert.assertEquals(cOutRels.size(), 2);
    Assert.assertEquals(cOutRels.get(0).getFromEntity(), "local:/sub/C-resolved.cdm.json/C");
    Assert.assertEquals(cOutRels.get(0).getToEntity(), "local:/B-resolved.cdm.json/B");
    Assert.assertEquals(cOutRels.get(1).getFromEntity(), "local:/sub/C-resolved.cdm.json/C");
    Assert.assertEquals(cOutRels.get(1).getToEntity(), "local:/sub/D-resolved.cdm.json/D");

    // D
    final CdmEntityDefinition dEnt = corpus.<CdmEntityDefinition>fetchObjectAsync(subManifest.getEntities().getAllItems().get(1).getEntityPath(), subManifest).join();
    final ArrayList<CdmE2ERelationship> dInRels = corpus.fetchIncomingRelationships(dEnt);
    final ArrayList<CdmE2ERelationship> dOutRels = corpus.fetchOutgoingRelationships(dEnt);
    Assert.assertEquals(dInRels.size(), 1);
    Assert.assertEquals(dInRels.get(0).getFromEntity(), "local:/sub/C-resolved.cdm.json/C");
    Assert.assertEquals(dInRels.get(0).getToEntity(), "local:/sub/D-resolved.cdm.json/D");
    Assert.assertEquals(dOutRels.size(), 0);

    // E
    final CdmEntityDefinition eEnt = corpus.<CdmEntityDefinition>fetchObjectAsync(resolvedManifest.getEntities().getAllItems().get(2).getEntityPath(), resolvedManifest).join();
    final ArrayList<CdmE2ERelationship> eInRels = corpus.fetchIncomingRelationships(eEnt);
    final ArrayList<CdmE2ERelationship> eOutRels = corpus.fetchOutgoingRelationships(eEnt);
    Assert.assertEquals(eInRels.size(), 1);
    Assert.assertEquals(eInRels.get(0).getFromEntity(), "local:/sub/F-resolved.cdm.json/F");
    Assert.assertEquals(eInRels.get(0).getToEntity(), "local:/E-resolved.cdm.json/E");
    Assert.assertEquals(eOutRels.size(), 0);

    // F
    final CdmEntityDefinition fEnt = corpus.<CdmEntityDefinition>fetchObjectAsync(subManifest.getEntities().getAllItems().get(2).getEntityPath(), subManifest).join();
    final ArrayList<CdmE2ERelationship> fInRels = corpus.fetchIncomingRelationships(fEnt);
    final ArrayList<CdmE2ERelationship> fOutRels = corpus.fetchOutgoingRelationships(fEnt);
    Assert.assertEquals(fInRels.size(), 0);
    Assert.assertEquals(fOutRels.size(), 1);
    Assert.assertEquals(fOutRels.get(0).getFromEntity(), "local:/sub/F-resolved.cdm.json/F");
    Assert.assertEquals(fOutRels.get(0).getToEntity(), "local:/E-resolved.cdm.json/E");
  }

  /**
   * Testing calculating relationships for the special kind of attribute that uses the "select one" directive
   */
  @Test
  public void testCalculateRelationshipsForSelectsOneAttribute()
      throws JsonMappingException, JsonProcessingException, IOException, InterruptedException, ExecutionException {
        final List<E2ERelationship> expectedRels = 
          JMapper.MAP.readValue(TestHelper.getExpectedOutputFileContent(
          TESTS_SUBPATH,
          "testCalculateRelationshipsForSelectsOneAttribute", 
          "expectedRels.json"),
          new TypeReference<List<E2ERelationship>>() {
        });
        final CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "testCalculateRelationshipsForSelectsOneAttribute", null);
        corpus.getStorage().mount("cdm", new LocalAdapter(TestHelper.SCHEMA_DOCS_ROOT));

        final CdmManifestDefinition manifest = corpus.<CdmManifestDefinition>fetchObjectAsync("local:/selectsOne.manifest.cdm.json").join();

        corpus.calculateEntityGraphAsync(manifest).get();
        manifest.populateManifestRelationshipsAsync().get();

        // check that each relationship has been created correctly
        verifyRelationships(manifest, expectedRels);
        
  }

  private static void verifyRelationships(CdmManifestDefinition manifest, List<E2ERelationship> expectedRelationships) {
    Assert.assertEquals(manifest.getRelationships().size(), expectedRelationships.size());

    for (final E2ERelationship expectedRel : expectedRelationships) {
      final List<CdmE2ERelationship> found = manifest.getRelationships().getAllItems()
        .parallelStream()
        .filter(x ->
          Objects.equals(expectedRel.getFromEntity(), x.getFromEntity())
          && Objects.equals(expectedRel.getFromEntityAttribute(), x.getFromEntityAttribute())
          && Objects.equals(expectedRel.getToEntity(), x.getToEntity())
          && Objects.equals(expectedRel.getToEntityAttribute(), x.getToEntityAttribute()))
        .collect(Collectors.toList());
      Assert.assertEquals(found.size(), 1);
   }
  }

  private static CdmManifestDefinition loadAndResolveManifest(CdmCorpusDefinition corpus, CdmManifestDefinition manifest, String renameSuffix) {
    System.out.println("Resolving manifest " + manifest.getManifestName() + " ...");
    CdmManifestDefinition resolvedManifest = manifest.createResolvedManifestAsync(manifest.getManifestName() + renameSuffix, "{n}-resolved.cdm.json").join();
    for (final CdmManifestDeclarationDefinition subManifestDecl : manifest.getSubManifests()) {
      final CdmManifestDefinition subManifest = corpus.<CdmManifestDefinition>fetchObjectAsync(subManifestDecl.getDefinition(), manifest).join();
      final CdmManifestDefinition resolvedSubManifest = loadAndResolveManifest(corpus, subManifest, renameSuffix);

      final CdmManifestDeclarationDefinition resolvedDecl = corpus.<CdmManifestDeclarationDefinition>makeObject(CdmObjectType.ManifestDeclarationDef, resolvedSubManifest.getManifestName());
      resolvedDecl.setDefinition(corpus.getStorage().createRelativeCorpusPath(resolvedSubManifest.getAtCorpusPath(), resolvedManifest));

      resolvedManifest.getSubManifests().add(resolvedDecl);
    }
    return resolvedManifest;
  }
}

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel.Tests/Cdm/Relationship/RelationshipTest.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Tests.Cdm
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Storage;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.Tools.Processor;
    using Microsoft.VisualStudio.TestTools.UnitTesting;
    using Newtonsoft.Json.Linq;
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Threading.Tasks;

    [TestClass]
    public class RelationshipTest
    {
        /// <summary>
        /// The path between TestDataPath and TestName.
        /// </summary>
        private string testsSubpath = Path.Combine("Cdm", "Relationship");

        /// <summary>
        /// Testing calculation of relationships and that those relationships are
        /// properly added to manifest objects
        /// </summary>
        [TestMethod]
        public async Task TestCalculateRelationshipsAndPopulateManifests()
        {
            var corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests");

            var rootManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>("local:/default.manifest.cdm.json");
            var subManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>(rootManifest.SubManifests[0].Definition);

            await corpus.CalculateEntityGraphAsync(rootManifest);
            await rootManifest.PopulateManifestRelationshipsAsync();

            var expectedAllManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests", "expectedAllManifestRels.json")).ToObject<List<E2ERelationship>>();
            var expectedAllSubManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests", "expectedAllSubManifestRels.json")).ToObject<List<E2ERelationship>>();

            // check that each relationship has been created correctly
            VerifyRelationships(rootManifest, expectedAllManifestRels);
            VerifyRelationships(subManifest, expectedAllSubManifestRels);
        }

        /// <summary>
        /// Testing calculation of relationships and that those relationships are
        /// properly added to manifest objects setting the populate flag to Exclusive
        /// </summary>
        [TestMethod]
        public async Task TestCalculateRelationshipsAndPopulateManifestsWithExclusiveFlag()
        {
            var corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests");

            var rootManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>("local:/default.manifest.cdm.json");
            var subManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>(rootManifest.SubManifests[0].Definition);

            await corpus.CalculateEntityGraphAsync(rootManifest);
            // make sure only relationships where to and from entities are in the manifest are found with the "exclusive" option is passed in
            await rootManifest.PopulateManifestRelationshipsAsync(CdmRelationshipDiscoveryStyle.Exclusive);

            var expectedExclusiveManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests", "expectedExclusiveManifestRels.json")).ToObject<List<E2ERelationship>>();
            var expectedExclusiveSubManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests", "expectedExclusiveSubManifestRels.json")).ToObject<List<E2ERelationship>>();

            // check that each relationship has been created correctly
            VerifyRelationships(rootManifest, expectedExclusiveManifestRels);
            VerifyRelationships(subManifest, expectedExclusiveSubManifestRels);
        }

        /// <summary>
        /// Testing calculation of relationships and that those relationships are
        /// properly added to manifest objects setting the populate flag to None
        /// </summary>
        [TestMethod]
        public async Task TestCalculateRelationshipsAndPopulateManifestsWithNoneFlag()
        {
            var corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestCalculateRelationshipsAndPopulateManifests", null);

            var rootManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>("local:/default.manifest.cdm.json");
            var subManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>(rootManifest.SubManifests[0].Definition);

            await corpus.CalculateEntityGraphAsync(rootManifest);
            // make sure no relationships are added when "none" relationship option is passed in
            await rootManifest.PopulateManifestRelationshipsAsync(CdmRelationshipDiscoveryStyle.None);

            Assert.AreEqual(0, rootManifest.Relationships.Count);
            Assert.AreEqual(0, subManifest.Relationships.Count);
        }

        /// <summary>
        /// Testing calculation of relationships and that those relationships are
        /// properly added to manifest objects
        /// </summary>
        [TestMethod]
        public async Task TestCalculateRelationshipsOnResolvedEntities()
        {
            var expectedResolvedManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsOnResolvedEntities", "expectedResolvedManifestRels.json")).ToObject<List<E2ERelationship>>();
            var expectedResolvedSubManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsOnResolvedEntities", "expectedResolvedSubManifestRels.json")).ToObject<List<E2ERelationship>>();
            var expectedResolvedExcManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsOnResolvedEntities", "expectedResolvedExcManifestRels.json")).ToObject<List<E2ERelationship>>();
            var expectedResolvedExcSubManifestRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsOnResolvedEntities", "expectedResolvedExcSubManifestRels.json")).ToObject<List<E2ERelationship>>();

            CdmCorpusDefinition corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestCalculateRelationshipsOnResolvedEntities");

            var rootManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>("local:/default.manifest.cdm.json");

            var resolvedManifest = await LoadAndResolveManifest(corpus, rootManifest, "-resolved");
            string subManifestPath = corpus.Storage.CreateAbsoluteCorpusPath(resolvedManifest.SubManifests[0].Definition);
            CdmManifestDefinition subManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>(subManifestPath) as CdmManifestDefinition;

            // using createResolvedManifest will only populate exclusive relationships
            VerifyRelationships(resolvedManifest, expectedResolvedExcManifestRels);
            VerifyRelationships(subManifest, expectedResolvedExcSubManifestRels);

            // check that each relationship has been created correctly with the all flag
            await resolvedManifest.PopulateManifestRelationshipsAsync();
            await subManifest.PopulateManifestRelationshipsAsync();
            VerifyRelationships(resolvedManifest, expectedResolvedManifestRels);
            VerifyRelationships(subManifest, expectedResolvedSubManifestRels);

            // it is not enough to check if the relationships are correct.
            // We need to check if the incoming and outgoing relationships are
            // correct as well. One being correct can cover up the other being wrong

            // A
            var aEnt = await corpus.FetchObjectAsync<CdmEntityDefinition>(resolvedManifest.Entities[0].EntityPath, resolvedManifest);
            var aInRels = corpus.FetchIncomingRelationships(aEnt);
            var aOutRels = corpus.FetchOutgoingRelationships(aEnt);
            Assert.AreEqual(0, aInRels.Count);
            Assert.AreEqual(1, aOutRels.Count);
            Assert.AreEqual("local:/A-resolved.cdm.json/A", aOutRels[0].FromEntity);
            Assert.AreEqual("local:/B-resolved.cdm.json/B", aOutRels[0].ToEntity);

            // B
            var bEnt = await corpus.FetchObjectAsync<CdmEntityDefinition>(resolvedManifest.Entities[1].EntityPath, resolvedManifest);
            var bInRels = corpus.FetchIncomingRelationships(bEnt);
            var bOutRels = corpus.FetchOutgoingRelationships(bEnt);
            Assert.AreEqual(2, bInRels.Count);
            Assert.AreEqual("local:/A-resolved.cdm.json/A", bInRels[0].FromEntity);
            Assert.AreEqual("local:/B-resolved.cdm.json/B", bInRels[0].ToEntity);
            Assert.AreEqual("local:/sub/C-resolved.cdm.json/C", bInRels[1].FromEntity);
            Assert.AreEqual("local:/B-resolved.cdm.json/B", bInRels[1].ToEntity);
            Assert.AreEqual(0, bOutRels.Count);

            // C
            var cEnt = await corpus.FetchObjectAsync<CdmEntityDefinition>(subManifest.Entities[0].EntityPath, subManifest);
            var cInRels = corpus.FetchIncomingRelationships(cEnt);
            var cOutRels = corpus.FetchOutgoingRelationships(cEnt);
            Assert.AreEqual(0, cInRels.Count);
            Assert.AreEqual(2, cOutRels.Count);
            Assert.AreEqual("local:/sub/C-resolved.cdm.json/C", cOutRels[0].FromEntity);
            Assert.AreEqual("local:/B-resolved.cdm.json/B", cOutRels[0].ToEntity);
            Assert.AreEqual("local:/sub/C-resolved.cdm.json/C", cOutRels[1].FromEntity);
            Assert.AreEqual("local:/sub/D-resolved.cdm.json/D", cOutRels[1].ToEntity);

            // D
            var dEnt = await corpus.FetchObjectAsync<CdmEntityDefinition>(subManifest.Entities[1].EntityPath, subManifest);
            var dInRels = corpus.FetchIncomingRelationships(dEnt);
            var dOutRels = corpus.FetchOutgoingRelationships(dEnt);
            Assert.AreEqual(1, dInRels.Count);
            Assert.AreEqual("local:/sub/C-resolved.cdm.json/C", dInRels[0].FromEntity);
            Assert.AreEqual("local:/sub/D-resolved.cdm.json/D", dInRels[0].ToEntity);
            Assert.AreEqual(0, dOutRels.Count);

            // E
            var eEnt = await corpus.FetchObjectAsync<CdmEntityDefinition>(resolvedManifest.Entities[2].EntityPath, resolvedManifest);
            var eInRels = corpus.FetchIncomingRelationships(eEnt);
            var eOutRels = corpus.FetchOutgoingRelationships(eEnt);
            Assert.AreEqual(1, eInRels.Count);
            Assert.AreEqual(0, eOutRels.Count);
            Assert.AreEqual("local:/sub/F-resolved.cdm.json/F", eInRels[0].FromEntity);
            Assert.AreEqual("local:/E-resolved.cdm.json/E", eInRels[0].ToEntity);

            // F
            var fEnt = await corpus.FetchObjectAsync<CdmEntityDefinition>(subManifest.Entities[2].EntityPath, subManifest);
            var fInRels = corpus.FetchIncomingRelationships(fEnt);
            var fOutRels = corpus.FetchOutgoingRelationships(fEnt);
            Assert.AreEqual(0, fInRels.Count);
            Assert.AreEqual(1, fOutRels.Count);
            Assert.AreEqual("local:/sub/F-resolved.cdm.json/F", fOutRels[0].FromEntity);
            Assert.AreEqual("local:/E-resolved.cdm.json/E", fOutRels[0].ToEntity);
        }

        /// <summary>
        /// Testing calculating relationships for the special kind of attribute that uses the "select one" directive
        /// </summary>
        [TestMethod]
        public async Task TestCalculateRelationshipsForSelectsOneAttribute()
        {
            var expectedRels = JToken.Parse(TestHelper.GetExpectedOutputFileContent(testsSubpath, "TestCalculateRelationshipsForSelectsOneAttribute", "expectedRels.json")).ToObject<List<E2ERelationship>>();
            var corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestCalculateRelationshipsForSelectsOneAttribute");
            corpus.Storage.Mount("cdm", new LocalAdapter(TestHelper.SchemaDocumentsPath));

            var manifest = await corpus.FetchObjectAsync<CdmManifestDefinition>("local:/selectsOne.manifest.cdm.json");

            await corpus.CalculateEntityGraphAsync(manifest);
            await manifest.PopulateManifestRelationshipsAsync();

            // check that each relationship has been created correctly
            VerifyRelationships(manifest, expectedRels);
        }

        private static void VerifyRelationships(CdmManifestDefinition manifest, List<E2ERelationship> expectedRelationships)
        {
            Assert.AreEqual(manifest.Relationships.Count, expectedRelationships.Count);

            foreach (E2ERelationship expectedRel in expectedRelationships)
            {
                List<CdmE2ERelationship> found = manifest.Relationships.Where(x =>
                x.FromEntity == expectedRel.FromEntity
                && x.FromEntityAttribute == expectedRel.FromEntityAttribute
                && x.ToEntity == expectedRel.ToEntity
                && x.ToEntityAttribute == expectedRel.ToEntityAttribute
                ).ToList();
                Assert.AreEqual(1, found.Count);
            }
        }

        private async static Task<CdmManifestDefinition> LoadAndResolveManifest(CdmCorpusDefinition corpus, CdmManifestDefinition manifest, string renameSuffix)
        {
            Console.WriteLine("Resolving manifest " + manifest.ManifestName + " ...");
            CdmManifestDefinition resolvedManifest = await manifest.CreateResolvedManifestAsync(manifest.ManifestName + renameSuffix, "{n}-resolved.cdm.json");

            foreach (CdmManifestDeclarationDefinition subManifestDecl in manifest.SubManifests)
            {
                CdmManifestDefinition subManifest = await corpus.FetchObjectAsync<CdmManifestDefinition>(subManifestDecl.Definition, manifest);
                CdmManifestDefinition resolvedSubManifest = await LoadAndResolveManifest(corpus, subManifest, renameSuffix);

                CdmManifestDeclarationDefinition resolvedDecl = corpus.MakeObject<CdmManifestDeclarationDefinition>(CdmObjectType.ManifestDeclarationDef, resolvedSubManifest.ManifestName);
                resolvedDecl.Definition = corpus.Storage.CreateRelativeCorpusPath(resolvedSubManifest.AtCorpusPath, resolvedManifest);

                resolvedManifest.SubManifests.Add(resolvedDecl);
            }

            return resolvedManifest;
        }
    }
}
